### PR TITLE
repart: Add DefaultSubvolume= setting

### PR DESCRIPTION
**Pull request by daandemeyer**
We already have Subvolumes= to create subvolumes, let's add DefaultSubvolume= as well to set the default subvolume.

### CODE REVIEW

File: src/partition/repart.c [hunk 1]
```diff
@@ -3736,6 +3782,10 @@ static int prepare_temporary_file(PartitionTarget *t, uint64_t size) {
         return 0;
 }
 
+static bool loop_device_error_is_fatal(const Partition *p, int r) {
+        return arg_offline == 0 || (r != -ENOENT && !ERRNO_IS_PRIVILEGE(r)) || !strv_isempty(p->subvolumes) || p->default_subvolume;
```
Comment (bluca) [inline src/partition/repart.c:3786]: assert(p)

File: src/partition/repart.c [hunk 2]
```diff
@@ -1668,6 +1669,42 @@ static int config_parse_make_dirs(
...
+                log_syntax(unit, LOG_WARNING, filename, line, r,
+                           "Failed to expand specifiers in DefaultSubvolume= parameter, ignoring: %s", rvalue);
+                return 0;
+        }
+
+        r = path_simplify_and_warn(p, PATH_CHECK_ABSOLUTE, unit, filename, line, lvalue);
+        if (r < 0)
+                return 0;
+
+        *subvol = TAKE_PTR(p);
```
Comment (YHNdnzj) [inline src/partition/repart.c]: free_and_replace?

File: src/shared/btrfs-util.c [hunk 3]
```diff
@@ -1785,6 +1785,22 @@ int btrfs_subvol_auto_qgroup(const char *path, uint64_t subvol_id, bool create_i
         return btrfs_subvol_auto_qgroup_fd(fd, subvol_id, create_intermediary_qgroup);
 }
 
+int btrfs_subvol_make_default(const char *path) {
+        _cleanup_close_ int fd = -EBADF;
+        uint64_t id;
+        int r;
+
```
Comment (YHNdnzj) [inline src/shared/btrfs-util.c:1792]: assert(path)

File: src/shared/btrfs-util.c [hunk 4]
```diff
@@ -1785,6 +1785,22 @@ int btrfs_subvol_auto_qgroup(const char *path, uint64_t subvol_id, bool create_i
...
+        fd = open(path, O_RDONLY|O_NOCTTY|O_CLOEXEC|O_DIRECTORY);
```
Comment (YHNdnzj) [inline src/shared/btrfs-util.c]: We usually don't specify O_RDONLY when O_DIRECTORY is used?

File: src/partition/repart.c [hunk 5]
```diff
@@ -4878,6 +4929,27 @@ static int do_make_directories(Partition *p, const char *root) {
...
+        _cleanup_free_ char *path = NULL;
+        int r;
+
+        assert(p);
+        assert(root);
+
+        if (!p->default_subvolume)
+                return 0;
+
+        path = path_join(root, p->default_subvolume, NULL);
```
Comment (YHNdnzj) [inline src/partition/repart.c]:
```suggestion
        path = path_join(root, p->default_subvolume);
```

File: man/repart.d.xml [hunk 6]
```diff
@@ -551,6 +551,22 @@
         <xi:include href="version-info.xml" xpointer="v255"/></listitem>
       </varlistentry>
 
+      <varlistentry>
+        <term><varname>DefaultSubvolume=</varname></term>
+
+        <listitem><para>Takes an absolute path specifying the default subvolume within the new filesystem.
+        Note that this setting does not create the subvolume itself, that can be configured with
+        <varname>Subvolumes=</varname>.</para>
```
Comment (poettering) [inline man/repart.d.xml:559]: why not just imply an appropriate Subvolumes= line if DefaultSubvolume= is not used?

**Comment by DaanDeMeyer** [comment]
@poettering One of our OpenSUSE friends from the image summit was playing around with systemd-repart and they needed this to replicate their subvolume layout.

File: man/repart.d.xml (hunk 6 above)
Comment (daandemeyer) [reply man/repart.d.xml:559]: Ehh we didn't do that for `Subvolumes=` either (which requires `MakeDirectories=` or `CopyFiles=`), so I just did the same for this to be consistent.

**Pull request merged**